client_side/.infrastructure/infrastructure.sh

```diff
@@ -0,0 +1,201 @@
+# Saves the user's current working directory and returnst to it once the
+# directory is set up.
+make_fs() {
+  pushd "${INFRA_DIR}" &> /dev/null
+
+  rm -rf "${FS_DIR}"
+  mkdir "${FS_DIR}" > /dev/null
+  tar -xzf "${INFRA_DIR}/fs.tgz" -C "${FS_DIR}"
+
+  # Moves the user to this directory
+  popd &> /dev/null
+}
+
+# Prints the list of resources that the user is allowed to use based on the
+# current treatment
+print_treatment() {
+  echo -n "For this half of the experiment you can use any online resources," \
+    "man pages, "
+  if [[ "$(cat "${INFRA_DIR}/.treatment")" == "T" ]]; then
+    echo "and Tellina <URL> to help you solve the tasks."
+  else
+    echo "but you can't use Tellina to help you solve the tasks."
+  fi
+}
+
+# Prints the current task number and its description.
+print_task() {
+  local task_code=$(cat "${INFRA_DIR}/.task_code")
+
+  echo "-----------------------------------------------------------------------"
+  echo "Task: ${curr_task}/${TASKS_SIZE}"
+
+  jq -r '.description' "${TASKS_DIR}/task_${task_code}/task_${task_code}.json"
+}
+
+# See documentation for ./verify_task.py for more details on what it does
+# Captures the stdout of verify_task.py into $status
+# Captures the exit code of verify_task.py into $EXIT
+verify_task() {
+  # Verify the output of the previous command.
+  local task_code="$(cat "${INFRA_DIR}/.task_code")"
+  local user_command="$(cat "${INFRA_DIR}/.command")"
+
+  status=$("${INFRA_DIR}"/verify_task.py ${task_code} ${user_command})
+  EXIT=$?
+}
+
+# Increment the current task number and update the true task code in .task_code
+# accordingly
+#   - Check if all the tasks are complete, if it is then skip the following
+#     steps and clean up instead.
+#   - If this current_task is equal to `TASKS_SIZE / 2`, switch treatments and
+#     task sets and notify the user.
+# Write "start task" to `.command`.
+# Writes to the log
+# Set time_elapsed=0, status="incomplete", SECONDS=0
+# Prints out the task's information
+next_task() {
+  make_fs
+
+  # Increment the number of tasks finished by the user
+  curr_task=$(( curr_task + 1 ))
+  echo "${curr_task}" > "${INFRA_DIR}/.curr_task"
+
+  # If we're done with all the tasks
+  if (( curr_task == TASKS_SIZE )); then
+    end_experiment
+
+    return 1
+  fi
+
+  # otherwise we check if we need to switch the task set and the treatment
+  # And go into the second half of the experiment
+  if (( curr_task == TASKS_SIZE / 2 + 1 )); then
+    echo "You have finished the first half of the experiment!"
+
+    # Updates the treatment and the task set for the user
+    if [[ "$(cat "${INFRA_DIR}/.treatment")" == "T" ]]; then
+      echo "N" > "${INFRA_DIR}/.treatment"
+    else
+      echo "T" > "${INFRA_DIR}/.treatment"
+    fi
+    if ((task_set == 1)); then
+      task_set=2
+    else
+      task_set=1
+    fi
+
+    print_treatment
+  fi
+
+  echo $(get_task_code) > "${INFRA_DIR}/.task_code"
```

Comment:
From here down should be abstracted out into a start_task function.
Maybe it should also incorporate most of the logic above and handle switching the treatment too (or calling a separate function which does so).  Note that switching the treatment may require more training.